pkg/ingester-rf1/metastore/metastore_hack.go

```diff
@@ -72,3 +78,15 @@ func (m *metastoreState) applyTruncate(request *raftlogpb.TruncateCommand) (*any
 	}
 	return &anypb.Any{}, nil
 }
+
+func (m *metastoreState) deleteSegment(ctx context.Context, bucket *bbolt.Bucket, segment *metastorepb.BlockMeta) error {
+	ctx, cancelFunc := context.WithTimeout(ctx, 15*time.Second)
+	defer cancelFunc()
+	if err := m.store.DeleteObject(ctx, path.Join(wal.Dir, segment.Id)); err != nil && !m.store.IsObjectNotFoundErr(err) {
+		return fmt.Errorf("failed to delete segment from object store: %w", err)
+	}
+	if err := bucket.Delete([]byte(segment.Id)); err != nil {
```

Comment:
We should re-order these - if the store delete completes but the metastore delete fails then we will error out when querying. If it happens the other way around, we'll have a few stray files that the lifecycle rule can clean up later.

Reply:
Hmm. I ordered this way intentionally so we wouldn't have orphaned files, and the operation would be "committed" when the command is applied to the Raft state machine. Why wouldn't we ignore missing files on querying?

Reply:
If we fail to delete the metadata after deleting the actual data, a future query will get the metadata entry but there won't be a file in object storage to fetch. When I tested this locally, the whole query fails in the case where it can't fetch a needed index - which sort of makes sense because we can't guarantee we're returning complete results otherwise. 
Though I could see it both ways - Maybe it makes more sense to alert on a "block missing" metric but return best-effort results to the user when querying?

Reply:
Even if I re-order these, the problem still exists, because one node will delete the objects from object storage before the command is replicated on the others. That means the other two replicas (assuming 3 total) will have out of date boltdb databases and in-memory caches.